File: python/day4/main.py
import re
import aoc


input = aoc.get_input('day4.txt')

valid = 0
valid2 = 0
details = {}

def num_in_range(s, min, max, digits=None):
    if digits is not None and len(s) != digits: return False
    try:
        i = int(s)
    except:
        return False
    if i < min or i > max: return False
    return True

def is_valid(details):
    if not num_in_range(details['byr'], 1920, 2002, 4): return False
    if not num_in_range(details['iyr'], 2010, 2020, 4): return False
    if not num_in_range(details['eyr'], 2020, 2030, 4): return False

    hgt = details['hgt']
    if hgt.endswith('cm'):
        if not num_in_range(hgt[:-2], 150, 193): return False
    elif hgt.endswith('in'):
        if not num_in_range(hgt[:-2], 59, 76): return False
    else:
        return False

    hcl = details['hcl']
    if not re.match('^#[0-9a-f]{6}$', hcl): return False

    if not details['ecl'] in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']: return False

    if not re.match('^[0-9]{9}$', details['pid']): return False

    return True

for line in input.splitlines() + ['']:
    if not line:
        if {'byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'} <= set(details.keys()):
            valid += 1
            valid2 += is_valid(details)
        details = {}

    entries = line.split()
    for entry in entries:
        key, value = entry.split(':')
        details[key] = value

print('Part 1:', valid)
print('Part 2:', valid2)

File: ruby/day4/main.rb
require 'set'
require_relative '../aoc'

def num_in_range(str, min, max, digits: nil)
  return false if !digits.nil? && str.length != digits
  Integer(str).between?(min, max)
ensure
  false
end

Passport = Struct.new(:byr, :iyr, :eyr, :hgt, :hcl, :ecl, :pid, keyword_init: true) do
  def valid?
    num_in_range(byr, 1920, 2002, digits: 4) &&
      num_in_range(iyr, 2010, 2020, digits: 4) &&
      num_in_range(eyr, 2020, 2030, digits: 4) &&
      (hgt.end_with?('cm') && num_in_range(hgt[...-2], 150, 193) ||
        hgt.end_with?('in') && num_in_range(hgt[...-2], 59, 76)) &&
      /^#[0-9a-f]{6}$/.match?(hcl) &&
      Set['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'].include?(ecl) &&
      /^[0-9]{9}$/.match?(pid)
  end
end

s = AOC::Solution.new(4)

s.preprocess do |input|
  required_keys = Passport.members.map(&:to_s).to_set

  input.split("\n\n").filter_map do |group_str|
    details = group_str.split.map {_1.split(':')}.to_h
    details.delete('cid')
    Passport.new(details) if details.keys.map.to_set == required_keys
  end
end

s.part1 {_1.length}
s.part2 {_1.count(&:valid?)}